# Section 1: Numerical Data Types & Arithmetic Operators ( The Road Trip Planner)

In [ ]:
''' A. How many liters of fuel will you need for the entire trip?'''

# Given values
fuel_efficiency = 11.50  # km per liter
total_distance = 475  # km
fuel_price = 1.65  # dollars per liter
tank_capacity = 50  # liters

# Fuel needed for the trip
fuel_needed = total_distance / fuel_efficiency
print("Fuel needed (liters):", round(fuel_needed, 2))

Fuel needed (liters): 41.3


Section 1.A

To find the total liters of fuel needed, we divide the total distance (475 km) by the car's fuel efficiency (11.50 km per liter). This gives the exact amount of fuel required to cover the trip.

In [ ]:
''' B. How many full tanks of fuel does this represent?'''

# Number of full tanks required (using floor division)
full_tanks = int(fuel_needed // tank_capacity)
print("Full tanks needed:", full_tanks)

Full tanks needed: 0


Section 1.B

We use floor division // to find how many full 50-liter tanks are required. Since, 41.30 liters is less than one full tank (50 liters), you need 0 additional full tanks beyond your starting full tank. Also, The // operator gives the whole number of times 50 liters fits into the fuel needed.

In [ ]:
''' C. What is the total estimated fuel cost for the trip?'''

# Total fuel cost
total_cost = fuel_needed * fuel_price
print("Total fuel cost ($):", round(total_cost, 2))

Total fuel cost ($): 68.15


Section 1.C

We multiply the total liters of fuel needed (41.30) by the price per liter ($1.65) to get the total cost for the trip.

In [ ]:
''' D. How many liters of fuel will be remaining in your tank?'''

# Fuel remaining after trip
fuel_used_from_last_tank = fuel_needed % tank_capacity
fuel_remaining = tank_capacity - fuel_used_from_last_tank
print("Fuel remaining after trip (liters):", round(fuel_remaining, 2))

Fuel remaining after trip (liters): 8.7


Section 1.D

The fuel used from the last full tank is the remainder when dividing the total fuel needed by the tank capacity. Subtract this from the full tank capacity to find out how much fuel is left after the trip. This represents the leftover fuel in your tank after completing the journey, assuming you only refueled the exact amount needed.

# Section 2: Comparison & Logical Operators ( Theme Park Entry Rules )

In [ ]:
''' A. Boolean Expression & Evaluation'''

# Given variables
age = 30
is_weekday = True
is_family_day = False
group_size = 3

# Discount Eligibility
discount_eligibility = ((age <= 12) and is_weekday) or (age >= 65) or (is_family_day and group_size >= 4)
print("Discount eligibility:", discount_eligibility)

Discount eligibility: False


Section 2.A

The Boolean expression checks if a person qualifies for a discount by evaluating three conditions: whether the person is a child (age 12 or younger) on a weekday, whether they are a senior (age 65 or older), or whether it's a family day and they are in a group of 4 or more. In this case, the person is 30 years old, so they are not a child or a senior. Although it is a weekday, the child condition does not apply, and it is not a family day, and the group size is only 3, which is less than 4. Therefore, none of the conditions for a discount are met, and the final result of the expression is False, meaning the person is not eligible for a discount.

In [ ]:
''' B: New Values & Evaluation'''

# New variables
age = 10
is_weekday = True
is_family_day = False
group_size = 2

# Discount Eligibility
discount_eligibility = ((age <= 12) and is_weekday) or (age >= 65) or (is_family_day and group_size >= 4)
print("Discount eligibility:", discount_eligibility)

Discount eligibility: True


Section 2.B

In this updated value, we checks if a 10-year-old visiting on a weekday, who is not part of a Family Day group and is in a group of 2 people, qualifies for a theme park discount. The code evaluates three conditions: whether the person is a child on a weekday (which is true), whether they are a senior (which is false), and whether it's Family Day with a group of 4 or more (which is also false). Since the first condition is true and the expression uses or, the person qualifies for the discount. Therefore, the printed result is: Discount eligibility: True.

In [ ]:
# C: Explain step-by-step logic

''' Step-by-Step Evaluation
1. age = 10 → age <= 12 → True
2. is_weekday = True → So (age <= 12 and is_weekday) → True and True → True
3. age >= 65 → False
4. is_family_day = False and group_size = 2 → False and False → False'''


' Step-by-Step Evaluation\n1. age = 10 → age <= 12 → True\n2. is_weekday = True → So (age <= 12 and is_weekday) → True and True → True\n3. age >= 65 → False\n4. is_family_day = False and group_size = 2 → False and False → False'

Section 2.C

The discount is given if any of the three conditions are True. Here, the first condition is True, and the other two are False. The final evaluation is True or False or False, which is True.
So, the person qualifies for the discount.

# Section 3: String Data Types, Indexing & Slicing ( Message Decoder)

In [ ]:
''' A. What is the character at index 5 of secret_message?'''

secret_message = "PyThoN_Is_FuN_AnD_PoWeRfUl!"

# Character at index 5
char_at_5 = secret_message[5]
print("Character at index 5:", char_at_5)

Character at index 5: N


Section 3.A

In this queries, we defines a message stored in a variable called secret_message and uses indexing to get the character at position 5, which is 'N'. It stores that character in another variable named char_at_5 and then prints it. Since Python counts from 0, index 5 refers to the 6th character in the string.

In [ ]:
# B. Using positive indexing and slicing, extract the substring "Is_FuN".

# Extract "Is_FuN" using Positive Indexing
substring_positive = secret_message[7:13]
print("Substring using positive indexing:", substring_positive)

Substring using positive indexing: Is_FuN


Section 3.B

Using the slicing with [7:13] extracts characters starting from index 7 up to, but not including, index 13. This gives the substring 'Is_FuN', showing how to extract a specific part of a string using positive indices.

In [ ]:
# C. Using negative indexing and slicing, extract the substring "PoWeRfUl".

# Extract "PoWeRfUl" using Negative Indexing
substring_negative = secret_message[-9:-1]
print("Substring using negative indexing:", substring_negative)

Substring using negative indexing: PoWeRfUl


Section 3.C

Negative indexing counts from the end of the string, so -9 is the ninth character from the end, and -1 is the last character (not included). This slice returns 'PoWeRfUl', illustrating how to use negative indices for slicing.

In [ ]:
# D. Create a new string that consists of every 4th character of secret_message, starting from the first character (index 0)

secret_message = "PyThoN_Is_FuN_AnD_PoWeRfUl!"
every_fourth = secret_message[0::4]
print("Every 4th character:", every_fourth)

Every 4th character: PosNDWU


Section 3.D

In this question, the slice [0::4] start s at index 0 and takes every fourth character from the string. This is an example of slicing with a step to obtain regularly spaced characters.

In [ ]:
# E. What happens if you try to access secret_message[50]?

secret_message = "PyThoN_Is_FuN_AnD_PoWeRfUl!"
print(secret_message[50])

IndexError: string index out of range

Section 3.E

When I try to access secret_message, I get an IndexError because the string simply isn’t long enough to have a character at position 50. In Python, strings are indexed starting at 0, so the highest valid index is always one less than the length of the string. If I try to access an index that doesn’t exist-like 50 in this case-Python immediately lets me know with an error. This is Python’s way of protecting your program from trying to use data that isn’t there, and it helps you catch mistakes where you might have miscounted the length of your string.

# Section 4: Boolean Values & Expressions ( Predict and Verify )


In [ ]:
''' A. (x == z) and (y > x)'''

# Given variable
x = 15
y = 25
z = 15
text = 'adventure'

part_a = (x == z) and (y > x)
print("a.", part_a)

a. True


Section 4.A

This expression checks if x equals z and if y is greater than x. Since both conditions are True (15 == 15 and 25 > 15), the overall result is True. The and operator only returns True when both sides are True.

In [ ]:
''' B. (x != z) or (y < x) or (len(text) < 5)'''
text = "adventure"

part_b = (x != z) or (y < x) or (len(text) < 5)
print("b.", part_b)

b. False


Section 4.B

This expression uses the or operator, which returns True if any condition is True. However, all three conditions-x != z, y < x, and len(text) < 5-are False. Since none of the conditions are met, the result is False.

In [ ]:
''' C. not ((x > y) and (text == "Adventure"))'''

part_c = not ((x > y) and (text == "Adventure"))
print("c.", part_c)

c. True


Section 4.C

Here, the expression inside the parentheses is False because x > y is False and text == "Adventure" is also False (due to case sensitivity). The and operator requires both to be True, so it returns False. Applying not to False gives a final result of True.

# Section 5: String Properties & Basic Methods ( User Profile Setup )


In [ ]:
''' A. Attempt to Modify a String Character '''

username = "MyUser123"
username[0] = 'm'

# Output
''''str' object does not support item assignment'''

TypeError: 'str' object does not support item assignment

Section 5.A

Strings in Python are immutable, meaning their contents cannot be changed after creation. You cannot modify individual characters directly. To change the username, create a new string (e.g., username = 'm' + username[1:]).

In [ ]:
''' B. Quote Transformation '''

quote = " practice makes perfect "
formatted_quote = quote.strip().upper().replace("PERFECT", "PROGRESS")
print(formatted_quote)

Section 5.B

Explanation of Methods:

.strip(): Removes leading/trailing whitespace.
Turns " practice makes perfect " → "practice makes perfect".

.upper(): Converts the string to uppercase.
Turns "practice makes perfect" → "PRACTICE MAKES PERFECT".

.replace("PERFECT", "PROGRESS"): Replaces "PERFECT" with "PROGRESS".
Final result: "PRACTICE MAKES PROGRESS".

In [ ]:
''' C. String Lengths '''

original_length = len(quote)
formatted_length = len(formatted_quote)
print("Original length:", original_length)
print("Formatted length:", formatted_length)


Original length: 24
Formatted length: 23


Section 5.C

To find the length of a string in Python, I use the len() function. When I apply len(quote) to the original quote (" practice makes perfect "), it returns 24 because it counts every character, including spaces at the beginning and end. After the transformations, len(formatted_quote) returns 23, since the leading and trailing spaces have been removed and the word "PERFECT" has been replaced with the longer word "PROGRESS". This shows how string methods can change the total number of characters in a string.